airbyte-connector-builder-server/unit_tests/connector_builder/impl/test_default_api.py

```diff
@@ -267,6 +271,85 @@ def test_read_stream_with_logs():
+        pytest.param(None, 1, id="test_create_request_no_record_limit_n_records_exceed_max"),
+        pytest.param(1, 3, id="test_create_request_with_record_limit"),
+        pytest.param(3, 1, id="test_create_request_record_limit_exceeds_max"),
+    ],
+)
+def test_read_stream_record_limit(request_record_limit, max_record_limit):
+    request = {
+        "url": "https://demonslayers.com/api/v1/hashiras?era=taisho",
+        "headers": {"Content-Type": "application/json"},
+        "body": {"custom": "field"},
+    }
+    response = {"status_code": 200, "headers": {"field": "value"}, "body": '{"name": "field"}'}
+    mock_source_adapter = MagicMock()
+    mock_source_adapter.read_stream.return_value = iter([
+        request_log_message(request),
+        response_log_message(response),
+        record_message("hashiras", {"name": "Shinobu Kocho"}),
+        record_message("hashiras", {"name": "Muichiro Tokito"}),
+        request_log_message(request),
+        response_log_message(response),
+        record_message("hashiras", {"name": "Mitsuri Kanroji"}),
+        response_log_message(response),
+    ])
+    n_records = 2
+    if request_record_limit is None:
+        record_limit = max_record_limit
+    else:
+        record_limit = min(request_record_limit, max_record_limit)
+
+    with patch.object(DefaultApiImpl, "MAX_RECORD_LIMIT", max_record_limit):
+        with patch.object(DefaultApiImpl, "_create_low_code_adapter", return_value=mock_source_adapter):
+            api = DefaultApiImpl()
+            loop = asyncio.get_event_loop()
+            actual_response: StreamRead = loop.run_until_complete(
+                api.read_stream(
+                    StreamReadRequestBody(manifest=MANIFEST, config=CONFIG, stream="hashiras", record_limit=request_record_limit))
+            )
+            single_slice = actual_response.slices[0]
+            for i, actual_page in enumerate(single_slice.pages):
+                assert len(actual_page.records) == min([record_limit, n_records])
+
+
+def test_read_stream_limit_0():
+    request = {
+        "url": "https://demonslayers.com/api/v1/hashiras?era=taisho",
+        "headers": {"Content-Type": "application/json"},
+        "body": {"custom": "field"},
+    }
+    response = {"status_code": 200, "headers": {"field": "value"}, "body": '{"name": "field"}'}
+    mock_source_adapter = MagicMock()
+    mock_source_adapter.read_stream.return_value = iter([
+        request_log_message(request),
+        response_log_message(response),
+        record_message("hashiras", {"name": "Shinobu Kocho"}),
+        record_message("hashiras", {"name": "Muichiro Tokito"}),
+        request_log_message(request),
+        response_log_message(response),
+        record_message("hashiras", {"name": "Mitsuri Kanroji"}),
+        response_log_message(response),
+    ])
+    request_record_limit = 0
+
+    with patch.object(DefaultApiImpl, "_create_low_code_adapter", return_value=mock_source_adapter):
```

Comment:
same question, is this telling us that our code structure needs to change?

Reply:
Is there a particular structural change you had in mind? I imagine this function was patched out of convenience (it's used throughout the `test_default_api.py` file) - looks to me like there are multiple places downstream that we could patch instead if we wanted but that might be a little trickier. As a simple alternative, `_create_low_code_adapter` is a `staticmethod` and doesn't need to be a method on the class - it could be moved to its own function if we don't want to surface the details of `DefaultApiImpl` in tests.

Reply:
I think one restructuring that could make sense would be to move some implementation out of `DefaultApiImpl`, and have a separate class doing the business logic of fetching records. That separate class could take an `adapter` object, which could be injected during tests. Thoughts?

Reply:
I made a lighter-weight change than moving all of the record-fetching out of `DefaultApiImpl`, injecting the adapter class into `DefaultApiImpl.__init__` instead. This cleans up the patches.